std/bitmanip.d

```diff
@@ -799,6 +799,140 @@ private:
+    size_t[] source = [1, 2, 3, 3424234, 724398, 2309400104, 3094929400423];
+    auto si = size_t.sizeof * 8;
+    auto ba = BitArray(source, source.length * si);
+    foreach (n; 0 .. source.length * si)
+    {
+        auto nth_bit = cast(bool) (source[n / si] & (1L << (n % si)));
+        assert(ba[n] == nth_bit);
+    }
+    ---
+    The least significant bit in any `size_t` unit is the starting bit of this
+    unit, and the most significant bit is the last bit of this unit. Therefore,
+    passing e.g. an array of `int`s may result in a different `BitArray`
+    depending on the processor's endianness.
+
+    This constructor is the inverse of $(LREF opCast).
+
+    $(RED Warning: All unmapped bits in the final word will be set to 0.)
+
+    Params:
+        v = Source array. `v.length` must be a multple of `size_t.sizeof`.
+        numbits = Number of bits to be mapped from the source array, i.e.
+                  length of the created `BitArray`.
+    */
+    this(void[] v, size_t numbits) nothrow @nogc pure
+    in
+    {
+        assert(numbits <= v.length * 8);
+        assert(v.length % size_t.sizeof == 0);
+    }
+    do
+    {
+        _ptr = cast(size_t*) v.ptr;
+        _len = numbits;
+        if (endBits)
+        {
+            // Need to mask away extraneous bits from v.
+            _ptr[dim - 1] &= endMask;
+        }
+    }
+
+    ///
+    @system unittest
+    {
+        import std.algorithm.comparison : equal;
+
+        auto a = BitArray([1, 0, 0, 1, 1]);
+
+        // Inverse of the cast.
+        auto v = cast(void[]) a;
+        auto b = BitArray(v, a.length);
+
+        assert(b.length == 5);
+        assert(b.bitsSet.equal([0, 3, 4]));
+
+        // a and b share the underlying data.
+        a[0] = 0;
+        assert(b[0] == 0);
+        assert(a == b);
+    }
+
+    ///
+    @system unittest
+    {
+        import std.algorithm.comparison : equal;
+
+        size_t[] source = [0b1100, 0b0011];
+        auto ba = BitArray(source, source.length * size_t.sizeof * 8);
+        // The least significant bit in each unit is this unit's starting bit.
+        assert(ba.bitsSet.equal([2, 3, 64, 65]));
+    }
+
+    ///
+    @system unittest
+    {
+        size_t[] source = [1, 0b101, 3, 3424234, 724398, 2309400104, 3094929400423];
+        auto si = size_t.sizeof * 8;
+        auto ba = BitArray(source, source.length * si);
+        foreach (n; 0 .. source.length * si)
+        {
+            auto nth_bit = cast(bool) (source[n / si] & (1L << (n % si)));
+            assert(ba[n] == nth_bit);
+        }
+        // Mapping only part of the array.
+        import std.algorithm.comparison : equal;
+
+        auto bc = BitArray(source, si + 1);
+        assert(bc.bitsSet.equal([0, si]));
+        // The unmapped bits from the final word have been cleared.
+        assert(source[1] == 1);
+    }
+
+    // Deliberately undocumented: raw initialization of bit array.
+    this(size_t len, size_t* ptr) nothrow @nogc pure
```

Comment:
I'm wondering if this constructor should remain undocumented? I mean, it works almost in the same way as the one above taking `void[]`.